src/Poseidon/EntityTypes.hs

```diff
@@ -263,15 +264,35 @@ reportDuplicateIndividuals individuals = do -- loop over duplication groups
     duplicateGroup@(firstInd : _) <- filter ((>1) . length) . groupBy (\a b -> indInfoName a == indInfoName b) . sortOn indInfoName $ individuals
     return (firstInd, [SpecificInd n' (head g) p | IndividualInfo n' g p <- duplicateGroup])
 
+resolveEntityIndices :: (EntitySpec a) => Bool -> [a] -> IndividualInfoCollection -> [Int]
+resolveEntityIndices True  = resolveEntityIndicesOrdered
+resolveEntityIndices False = resolveEntityIndicesUnordered
+
 -- | this finds the indices of all individuals from an individual-list which are specified in the Entity list
-resolveEntityIndices :: (MonadThrow m, EntitySpec a) => [a] -> IndividualInfoCollection -> m [Int]
-resolveEntityIndices entities (indInfos, areLatest) = do
-    let relevantIndizes = [ i | (i, ind, l) <- zip3 [0..] indInfos areLatest, indInfoConformsToEntitySpecs ind l entities ]
-    return relevantIndizes
-
-resolveUniqueEntityIndices :: (EntitySpec a) => [a] -> IndividualInfoCollection -> PoseidonIO [Int]
-resolveUniqueEntityIndices entities indInfoCollection = do
-    relevantIndices <- resolveEntityIndices entities indInfoCollection
+resolveEntityIndicesUnordered :: (EntitySpec a) => [a] -> IndividualInfoCollection -> [Int]
+resolveEntityIndicesUnordered entities (indInfos, areLatest) =
+    [ i | (i, ind, l) <- zip3 [0..] indInfos areLatest, indInfoConformsToEntitySpecs ind l entities ]
+
+-- | this finds the indices of all individuals from an individual-list which are specified in the Entity list, ordered by the entity list
+resolveEntityIndicesOrdered :: (EntitySpec a) => [a] -> IndividualInfoCollection -> [Int]
+resolveEntityIndicesOrdered entities (indInfos, areLatest) = go [] entities
+  where
+    go :: (EntitySpec a) => [Int] -> [a] -> [Int]
+    go selectedIndices [] = selectedIndices
+    go selectedIndices (entity:restEntities)=
```

Comment:
Neat! I think this works. But I wonder whether things would become slightly more easy to read by handling the two paths separately:

```haskell
go :: (EntitySpec a) => [Int] -> [a] -> [Int]
go selectedIndices (entity:restEntities)
   | (isPoseidonEntity entity || isIncludeEntity) = ...
   | isExcludeEntity entity = ...
```

Maybe not. If we only had `SignedEntity` to consider, then it would be a simple pattern match, but with the `EntitySpec` class we need a case statement or guards.

Reply:
Yes, that was my original plan, but I realised then that `isExcludeEntity` is not a function that one can implement outside the class definition, because it needs to be implemented on a per-instance basis. In a polymorphic function like this one here, it is impossible to inquire about a sign-property without extending the class. I did not want to do this, so I decided for this more general implementation which always "cleans" the existing indices and adds additional ones. I think that's actually more elegant in a way because it is truly agnostic about signed or unsigned entities.